server/app/durablejobs/DurableJobRunner.java

```diff
@@ -0,0 +1,178 @@
+package durablejobs;
+
+import annotations.BindingAnnotations;
+import com.google.common.base.Preconditions;
+import com.typesafe.config.Config;
+import io.ebean.DB;
+import io.ebean.Database;
+import io.ebean.Transaction;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import models.PersistedDurableJob;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import repository.PersistedDurableJobRepository;
+
+/**
+ * Executes {@link DurableJob}s when their time has come.
+ *
+ * <p>The runner continues executing jobs as long as there are jobs to execute and it does not
+ * exceed the time specified by "durable_jobs.poll_interval_seconds". This is to prevent runners
+ * overlapping in the same server.
+ */
+public final class DurableJobRunner {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DurableJobRunner.class);
+
+  private final Database database = DB.getDefault();
+  private final DurableJobExecutionContext durableJobExecutionContext;
+  private final DurableJobRegistry durableJobRegistry;
+  private final int jobTimeoutMinutes;
+  private final PersistedDurableJobRepository persistedDurableJobRepository;
+  private final Provider<LocalDateTime> nowProvider;
+  private final ZoneOffset zoneOffset;
+  private final int runnerLifespanSeconds;
+
+  @Inject
+  public DurableJobRunner(
+      Config config,
+      DurableJobExecutionContext durableJobExecutionContext,
+      DurableJobRegistry durableJobRegistry,
+      PersistedDurableJobRepository persistedDurableJobRepository,
+      @BindingAnnotations.Now Provider<LocalDateTime> nowProvider,
+      ZoneId zoneId) {
+    this.durableJobExecutionContext = Preconditions.checkNotNull(durableJobExecutionContext);
+    this.durableJobRegistry = Preconditions.checkNotNull(durableJobRegistry);
+    this.jobTimeoutMinutes = config.getInt("durable_jobs.job_timeout_minutes");
+    this.persistedDurableJobRepository = Preconditions.checkNotNull(persistedDurableJobRepository);
+    this.runnerLifespanSeconds = config.getInt("durable_jobs.poll_interval_seconds");
+    this.nowProvider = Preconditions.checkNotNull(nowProvider);
+    this.zoneOffset = zoneId.getRules().getOffset(nowProvider.get());
+  }
+
```

Comment:
**suggestion:** Add javadoc for this method and maybe others?

Reply:
Done